ChiModules/KEigenvalueSolver/initialize_precursors.cc

```diff
@@ -15,45 +15,53 @@ using namespace LinearBoltzmann;
 /**Computes the point wise delayed neutron precursor concentrations.*/
 void KEigenvalue::Solver::InitializePrecursors()
 {
-  typedef SpatialDiscretization_PWLD  PWLD;
-  auto pwl = std::static_pointer_cast<PWLD>(discretization);
-
-  Nj_new_local.assign(Nj_new_local.size(), 0.0);
-
-  // ----- Loop over cells
-  for (auto& cell : grid->local_cells)
+  if (options.use_precursors)
   {
-    // ----- Cell information
-    const auto xs_id = matid_to_xs_map[cell.material_id];
-    auto& xs = material_xs[xs_id];
-    auto cell_fe_view = pwl->GetCellMappingFE(cell.local_id);
-    auto& transport_view = cell_transport_views[cell.local_id];
-
-    // ----- Loop over cell dofs
-    for (int i = 0; i < transport_view.NumNodes(); ++i)
+    precursor_new_local.assign(precursor_new_local.size(), 0.0);
+
+    //================================================== Loop over cells
+    for (auto& cell : grid->local_cells)
     {
-      size_t ir = transport_view.MapDOF(i, 0, 0);
-      size_t jr = pwl->MapDOFLocal(cell, i, Nj_unk_man, 0, 0);
-      double* Nj_newp  = &Nj_new_local[jr];
-      double* phi_newp = &phi_new_local[ir];
-
-      // ----- If a fissial material with precursors
-      if ((xs->is_fissile) and (xs->num_precursors > 0)) {
-        // ----- Loop over precursors
-        for (size_t j = 0; j < xs->num_precursors; ++j)
+      auto& full_cell_view = cell_transport_views[cell.local_id];
+
+      //==================== Obtain xs
+      int cell_matid = cell.material_id;
+      int xs_id = matid_to_xs_map[cell_matid];
+
+      if ((xs_id < 0) || (xs_id >= material_xs.size()))
+      {
+        chi_log.Log(LOG_ALLERROR)
+            << "Cross-section lookup error\n";
+        exit(EXIT_FAILURE);
+      }
+
+      auto xs = material_xs[xs_id];
+
+      //======================================== Loop over nodes
+      const int num_nodes = full_cell_view.NumNodes();
+      for (int i = 0; i < num_nodes; ++i)
+      {
+        size_t ir = full_cell_view.MapDOF(i, 0, 0);
+        size_t jr = discretization->MapDOFLocal(cell, i, precursor_uk_man, 0, 0);
+        double* Nj_newp = &precursor_new_local[jr];
```

Comment:
This should be removed and an offset used below.

Reply:
Agreed

Reply:
Fixed.